test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalUpdateTests.cs

```diff
@@ -0,0 +1,182 @@
+            var identifier = Guid.NewGuid().ToString();
+
+            observation.Identifier.Add(new Identifier("http://e2etests", identifier));
+            FhirResponse<Observation> response = await Client.CreateAsync(observation);
+            Assert.Equal(HttpStatusCode.Created, response.StatusCode);
+
+            var observation2 = Samples.GetDefaultObservation().ToPoco<Observation>();
+            observation2.Id = response.Resource.Id;
+            observation2.Identifier.Add(new Identifier("http://e2etests", identifier));
+            observation2.Text.Div = "<div>Updated!</div>";
+            FhirResponse<Observation> updateResponse = await Client.ConditionalUpdateAsync(
+                observation2,
+                $"identifier={identifier}");
+
+            Assert.Equal(HttpStatusCode.OK, updateResponse.StatusCode);
+
+            Observation updatedResource = updateResponse.Resource;
+
+            Assert.NotNull(updatedResource);
+            Assert.Equal(response.Resource.Id, updatedResource.Id);
+        }
+
+        [Fact]
+        [Trait(Traits.Priority, Priority.One)]
+        public async Task GivenAResourceWithIncorrectId_WhenUpsertingConditionallyWithOneMatch_TheServerShouldFail()
+        {
+            var observation = Samples.GetDefaultObservation().ToPoco<Observation>();
+            var identifier = Guid.NewGuid().ToString();
+
+            observation.Identifier.Add(new Identifier("http://e2etests", identifier));
+            FhirResponse<Observation> response = await Client.CreateAsync(observation);
+            Assert.Equal(HttpStatusCode.Created, response.StatusCode);
+
+            var observation2 = Samples.GetDefaultObservation().ToPoco<Observation>();
+            observation2.Id = Guid.NewGuid().ToString();
+
+            var exception = await Assert.ThrowsAsync<FhirException>(() => Client.ConditionalUpdateAsync(
+                observation2,
+                $"identifier={identifier}"));
+
+            Assert.Equal(HttpStatusCode.BadRequest, exception.Response.StatusCode);
+        }
+
+        [Fact]
+        [Trait(Traits.Priority, Priority.One)]
+        public async Task GivenAResource_WhenUpsertingConditionallyWithMultipleMatches_TheServerShouldFail()
+        {
+            var observation = Samples.GetDefaultObservation().ToPoco<Observation>();
+            var identifier = Guid.NewGuid().ToString();
+
+            observation.Identifier.Add(new Identifier("http://e2etests", identifier));
+
+            FhirResponse<Observation> response = await Client.CreateAsync(observation);
+            Assert.Equal(HttpStatusCode.Created, response.StatusCode);
+
+            FhirResponse<Observation> response2 = await Client.CreateAsync(observation);
+            Assert.Equal(HttpStatusCode.Created, response2.StatusCode);
+
+            var observation2 = Samples.GetDefaultObservation().ToPoco<Observation>();
+            observation2.Id = Guid.NewGuid().ToString();
```

Comment:
nit: We should set observation2.Id to either `null` or `response.Resource.Id`. Just to make sure we don't hit the IncorrectId flow (which the previous test tests for). The Assertion for the PreconditionFailed does make sure that we are hitting the right scenario, but just thought it would be cleaner to set the Id correctly.